Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,183 @@
+    ) async throws {
+        self.config = config
         self.topicConfig = topicConfig
+        self.logger = logger
+        self.topicHandles = [:]
+
+        self.referenceToSelf = Unmanaged.passRetained(self).toOpaque()
+        self.config.setDeliveryReportCallback(callback: self.deliveryReportCallback)
+        self.config.setCallbackOpaque(opaque: self.referenceToSelf)
+
+        self.client = try KafkaClient(type: .producer, config: self.config, logger: self.logger)
+        self.kafkaHandle = self.client.kafkaHandle
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Kafka producer must be closed **explicitly**.
+    public func shutdownGracefully() {
+        // Wait 10 seconds for outstanding messages to be sent and callbacks to be called
+        rd_kafka_flush(self.kafkaHandle, 10000)
+
+        for (_, topicHandle) in self.topicHandles {
+            rd_kafka_topic_destroy(topicHandle)
+        }
+        self.pollTask.cancel()
+
+        _ = Unmanaged<KafkaProducer>.fromOpaque(self.referenceToSelf).takeRetainedValue()
+    }
+
+    /// Send messages to the Kafka cluster asynchronously, aka "fire and forget".
+    /// This function is non-blocking.
+    /// - Parameter message: The ``KafkaProducerMessage`` that is sent to the KafkaCluster.
+    /// - Returns: Unique message identifier matching the `id` property of the corresponding ``KafkaAckedMessage``
+    @discardableResult
+    public func sendAsync(message: KafkaProducerMessage) throws -> UInt {
+        let topicHandle = self.createTopicHandleIfNeeded(topic: message.topic)
+
+        let keyBytes: [UInt8]?
+        if let key = message.key {
+            keyBytes = key.getBytes(at: 0, length: key.readableBytes)
+        } else {
+            keyBytes = nil
+        }
+
+        self.messageIDCounter += 1
+
+        let responseCode = message.value.withUnsafeReadableBytes { valueBuffer in
+
+            return rd_kafka_produce(
+                topicHandle,
+                message.partition.rawValue,
+                RD_KAFKA_MSG_F_COPY,
+                UnsafeMutableRawPointer(mutating: valueBuffer.baseAddress),
+                valueBuffer.count,
+                keyBytes,
+                keyBytes?.count ?? 0,
+                UnsafeMutableRawPointer(bitPattern: messageIDCounter)
+            )
+        }
+
+        guard responseCode == 0 else {
+            throw KafkaError(rawValue: responseCode)
+        }
+
+        return self.messageIDCounter
+    }
+
+    // Closure that is executed when a message has been acknowledged by Kafka
+    private let deliveryReportCallback: (
+        @convention(c) (OpaquePointer?, UnsafePointer<rd_kafka_message_t>?, UnsafeMutableRawPointer?) -> Void
+    ) = { _, messagePointer, producerPointer in
+
```

Comment:
I would prefer if we try to unwrap the `opaquePointer` (here called `producerPointer`) first. If it is `nil` we should `fatalError` because something in our code went wrong.

Once, we have the reference to `self` we should call a method on self that passes the `messagePointer`. This way we get off the C callback as fast as possible.

Inside this new method we should add least add a debug log when we get a callback for message and the `messagePointer` is `nil`.

Reply:
We also need to call [rd_kafka_message_destroy](https://docs.confluent.io/2.0.0/clients/librdkafka/rdkafka_8h.html#a35e0c519209d1afe9e29468f766c1c24) in this new method